### DISCUSSION TITLE
LESSON 11 - Can't see imported MoodNft's image in Metamask

### DISCUSSION THREAD

**Discussion by Wakushi** [Cyfrin/foundry-full-course-cu#462] [Q&A] [answered]
Hello everyone ! 
When I minted a MoodNft on an anvil chain, I was able to import the token in Metamask but instead of seeing the happy face, it's just a greyed box with "Untitled Collection #0" written inside. What could I have done wrong ? 

Here are my contract and deploy script : 

**Contract** 
```solidity 
//SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
import {Base64} from "@openzeppelin/contracts/utils/Base64.sol";
import {console} from "forge-std/Test.sol";

contract MoodNft is ERC721 {
    // ERRORS

    error MoodNft__CantFlipMoodIfNotOwner();

    // STATE VARIABLES

    uint256 private s_tokenCounter;
    string private s_sadSvgImageUri;
    string private s_happySvgImageUri;

    // ENUMS

    enum Mood {
        Sad,
        Happy
    }

    // MAPPINGS

    mapping(uint256 => Mood) s_tokenIdToMood;

    constructor(
        string memory sadSvgImageUri,
        string memory happySvgImageUri
    ) ERC721("MoodNft", "MN") {
        s_tokenCounter = 0;
        s_sadSvgImageUri = sadSvgImageUri;
        s_happySvgImageUri = happySvgImageUri;
    }

    function mintNft() public {
        _safeMint(msg.sender, s_tokenCounter);
        // Default to happy
        s_tokenIdToMood[s_tokenCounter] = Mood.Happy;
        s_tokenCounter++;
    }

    function flipMood(uint256 tokenId) public {
        // Only the owner can change the mood
        // ERC721 built in method to check if the sender is the owner of X token
        if (!_isApprovedOrOwner(msg.sender, tokenId)) {
            revert MoodNft__CantFlipMoodIfNotOwner();
        }
        if (s_tokenIdToMood[tokenId] == Mood.Happy) {
            s_tokenIdToMood[tokenId] = Mood.Sad;
        } else {
            s_tokenIdToMood[tokenId] = Mood.Happy;
        }
    }

    function _baseURI() internal pure override returns (string memory) {
        return "data:application/json;base64,";
    }

    function tokenURI(
        uint256 tokenId
    ) public view override returns (string memory) {
        string memory imageURI;
        if (s_tokenIdToMood[tokenId] == Mood.Happy) {
            imageURI = s_happySvgImageUri;
        } else {
            imageURI = s_sadSvgImageUri;
        }

        return
            // We make this all a string
            string(
                // We concatenate the baseURI with the imageURI so that it starts with data:application/...
                abi.encodePacked(
                    _baseURI(),
                    // We encode it in base 64
                    Base64.encode(
                        // We convert it to bytes to use the OpenZeppelin Base64 encoder
                        bytes(
                            // abi.encodePacked concatenates all the strings
                            abi.encodePacked(
                                '{ "name": ',
                                name(),
                                '", "description":"An NFT that reflects the owners mood.", "attributes": [{"trait_type": "moodiness", "value":100}], "image":"',
                                imageURI,
                                '"}'
                            )
                        )
                    )
                )
            );
    }
}
```

**Deploy Script**

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Script, console} from "forge-std/Script.sol";
import {MoodNft} from "../src/MoodNft.sol";
import {Base64} from "@openzeppelin/contracts/utils/Base64.sol";

contract DeployMoodNft is Script {
    function run() external returns (MoodNft) {
        string memory sadSvg = vm.readFile("./img/sad.svg");
        string memory happySvg = vm.readFile("./img/happy.svg");
        vm.startBroadcast();
        MoodNft moodnft = new MoodNft(
            svgToImageURI(sadSvg),
            svgToImageURI(happySvg)
        );
        vm.stopBroadcast();
        return moodnft;
    }

    function svgToImageURI(
        string memory svg
    ) public pure returns (string memory) {
        // Input : <svg width="1024px"....</svg>
        // Ouput : data:image/svg+xml;base64,PD94bWwgd...
        string memory baseURL = "data:image/svg+xml;base64,";
        string memory svgBase64Encoded = Base64.encode(
            bytes(string(abi.encodePacked(svg)))
        );
        return string(abi.encodePacked(baseURL, svgBase64Encoded));
    }
}

```

**Comment by Niferu** [comment]
Hello @Wakushi 

I'm not sure if you can import nft's from anvil blockchain as this chain is temporary and that is why you see grey box. To see proper NFT you need to mint it on Sepolia network for example and that for sure will be visible in your wallet after import.

Example from my wallet of custom NFT minted on Sepolia:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/57966272-6a5e-4e4e-a5dd-8befb32a47ca)

In case you will attach metadata with incorrect image source your NFT box will be also greyed out.

**Reply by Wakushi** [reply]
Hi @Neftyr !
That would indeed be logic, but in Patrick's video at [9:35:28](https://youtu.be/sas02qSFZ74?t=34528), we can see that he imports the MoodNFT in Metamask while being on the Anvil chain. 
Maybe they changed it afterwards ?

**Reply by Niferu** [reply, marked as answer]
@Wakushi 

In this case your url pointing to image from metadata is corrupted. I had similar problem and only cloning repo solved my issue as it was something wrong with formatting of svg files code (till today I have no idea what exactly coz code was the same but even copying it from repo didnt work I had to clone it).

**Reply by Wakushi** [reply]
Alright, thanks a lot, I'll try to clone as wall / deploy to Sepolia :)

**Reply by Niferu** [reply]
Let me know if it worked for you as SVG files can be tricky and feel free to ask any questions. I pretty much mastered NFT's topic by creating 2 big projects for hackathons, so I should be able to help :D

**Comment by robbiesumner** [comment]
I'd like to figure this out too. I have the same issue and spent about 2 hours trying to solve this and can't get it to work. Can we reopen this discussion.
I'm really interested!

**Reply by robbiesumner** [reply]
It doesn't seem to even render the URI, it doesn't show the description or the nft name either. However if I copy the string returned from `tokenURI(uint256 tokenId)` and paste it into my browser, I get the (seemingly) correct JSON object.

Maybe there is a minor error in the json object 🤷🏼‍♂️

**Comment by robbiesumner** [comment]
I found the issue for me and for you too!!!

It's just quotes that are missing in the JSON object. That way metamask can't decode the data and just shows no metadata. For you it is a double quotation mark right before `name()` thats missing.

Hope this makes sense now

**Reply by Csb303** [reply]
I have the same issue. How did you identify where the quotes are missing?

**Reply by robbiesumner** [reply]
It can be really hard sometimes to see what you're missing, you'll have to just stare at your code and compare it to others

You could also go to your AI fren and ask him, if you're missing any quotes or anything else:
I just tried it for the code from @Wakushi and it found it immediately: https://chat.openai.com/share/52706d67-f873-456f-8e1a-4c3dbe74fd85

**Reply by Csb303** [reply]
I have found this issue but still unsure what is causing it.

The "image" value in my JSON object begins with 'data:/image/svg=xml;base64' but for the NFT to display properly it needs to be formatted like this 'data:image/svg+xml;base64'
